BridgePool & BridgePoolFactory

gusjavaz opened the pull request:
## Scope
https://madhive.atlassian.net/browse/MP-269
https://madhive.atlassian.net/browse/MP-268
This PR includes merge of https://github.com/MadBase/MadNet/pull/126 with some changes

Commit:
Added first version of BridgePool with happy path

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into candidate

Commit:
Added new tests

Commit:
Merged name refactoring

Commit:
Add ubuntu support to scripts

Commit:
Added ubuntu support to scripts

Commit:
Merge branch 'candidate' of https://github.com/gusjavaz/MadNet into candidate

Commit:
Merge branch 'ubuntu-support' into candidate

Commit:
Added validator starting and first localrpc tests

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into add-localrpc-tests

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into add-localrpc-tests

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into add-localrpc-tests

Commit:
Added TX tests

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into add-localrpc-tests

kasandravtv committed:
Deposited events

Commit:
Code cleaning

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into add-localrpc-tests

Commit:
Implement MadByte refactoring changes

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into bridge-pool

kasandravtv committed:
add nonce checks

Commit:
Delete package-lock.json

Commit:
Added removed package-lock.json

Commit:
checkout package-lock.json from candidate

Commit:
Fixed golang-vet warnings

Commit:
Added withdraw function

Commit:
Added test to smt_test.go to create MerkleProof

Commit:
Fixed some conflicts

Commit:
Added deployment via factory and access control

Commit:
Added error label support

Commit:
Added processing of MerkleProof as binary

Commit:
Added missing fee when creating a transaction

Commit:
Changed library methods to receive MerkleProof struct

Commit:
Added integration with event emitter

Commit:
Fixed issues in tests

Commit:
Added extraction of stateRoot out of Snapshot

Commit:
Added param controls

Commit:
Fix conflicts

Commit:
Fix conflicts

Commit:
Merge candidate

Commit:
Added UML

Commit:
Code reorganized

Commit:
Fixed conflicts

Commit:
Prepare merge

Commit:
Commented out DepositNotifier references for future use

Commit:
Fixed lindt errors

Commit:
Fixed lindt errors

Commit:
Fixed lindt errors

Commit:
Fix problem in SnapshotsMock that was making ValidatorPool tests fail

Commit:
Fix problem in SnapshotsMock that was making ValidatorPool tests fail

Commit:
Added generic deployFactory for contracts other than AliceNet (and AliceNet)

Commit:
Added BridgePoolFactory Tests

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into bridge-pool-factory

Commit:
Revert BridgePoolFactory deployment as factory

Commit:
Revert minor change

Commit:
Fixed lindt errors

Commit:
Added support for error codes

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into candidate

Commit:
Merge remote-tracking branch 'kasvtv/events' into bridge-pool-factory-kasvtv-events

z-j-lin commented on bridge/contracts/BridgePool.sol:
```diff
@@ -0,0 +1,108 @@
+        _safeTransferERC20(IERC20Transferable(_erc20TokenContract), msg.sender, burnedUTXO.value);
+    }
+
+    receive() external payable {}
```
Locked eth. This contract needs a method to withdraw eth if it can receive eth.

gusjavaz replied:
Fixed in latest commit. Now the ETH obtained from BToken burning is immediately transferred to Foundation contract. Thanks @z-j-lin

gusjavaz replied:
According to the latest guidelines, the remaining ETH from the burning of fees is transferred to BToken through a selfdestructing contract.

Commit:
Added integration with DepositNotifier (PR-126)
Renamed DepositNotifier to BridgePoolDepositNotifier
ETH in BridgePool from BToken burning is now transfered to Foundation
Added method documentation in NatSpec format

Commit:
Added dist to .gitignore

Commit:
Merged .gitignore

Commit:
Added generated bindings for BridgePoolDepositNotifier

Commit:
Merged .gitignore

Commit:
Merge branch 'candidate' of https://github.com/MadBase/MadNet into candidate

Commit:
Fixed tests for BridgePoolDepositNotifier
Add error code contracts to fixture

Commit:
Recreate deploymentList and deploymentArgsTemplate with new contarcts

Commit:
Exclude  unrelated files from Kas branch

Commit:
Exclude one last unrelated file from Kasper's branch

Commit:
Removed package-lock.json

Commit:
Remove unrelated file

Commit:
Merged changes

Cr0wn-Gh0ul commented on .gitignore:
```diff
@@ -25,4 +26,14 @@ local-geth/*
 scripts/generated
-node_modules/
\ No newline at end of file
+localstate.swagger.json
```
Are we sure these need to be ignored?

gusjavaz replied:
Sorry, this is coming from merging with Kasper's PR branch (events). @kasvtv Are you ok to include this change in gitignore (and some others related to *gen.go) in this PR or you prefer to have them in your own?

gusjavaz replied:
Removed legacy gitignore lines on last commit

Commit:
Added tranfer to btoken of ETH from bToken burning

Commit:
Fixed transfer to eths from bToken Burning

Commit:
Merged ZJ development

Commit:
Added funds transfer through selfdestruct operation

Commit:
Added transfer burned fee to Btoken through temporary selfdestruct contract

Commit:
Removed finished TODOs

Commit:
Merge branch 'candidate' into candidate
